mmdet3d/models/losses/lovasz_loss.py

```diff
@@ -0,0 +1,365 @@
+    """
+    logits = logits.view(-1)
+    labels = labels.view(-1)
+    if ignore_index is None:
+        return logits, labels
+    valid = (labels != ignore_index)
+    vlogits = logits[valid]
+    vlabels = labels[valid]
+    return vlogits, vlabels
+
+
+def flatten_probs(
+        probs: torch.Tensor,
+        labels: torch.Tensor,
+        ignore_index: int = None) -> Tuple[torch.Tensor, torch.Tensor]:
+    """Flattens predictions and labels in the batch. Remove tensors whose
+    labels equal to 'ignore_index'.
+
+    Args:
+        probs (torch.Tensor): Predictions to be modified.
+        labels (torch.Tensor): Labels to be modified.
+        ignore_index (int | None): The label index to be ignored.
+            Defaults to None.
+
+    Return:
+        tuple(torch.Tensor, torch.Tensor): Modified predictions and labels.
+    """
+    if probs.dim() == 3:
+        # assumes output of a sigmoid layer
+        B, H, W = probs.size()
+        probs = probs.view(B, 1, H, W)
+    B, C, H, W = probs.size()
+    probs = probs.permute(0, 2, 3, 1).contiguous().view(-1, C)  # B*H*W, C=P,C
+    labels = labels.view(-1)
+    if ignore_index is None:
+        return probs, labels
+    valid = (labels != ignore_index)
+    vprobs = probs[valid.nonzero().squeeze()]
+    vlabels = labels[valid]
+    return vprobs, vlabels
+
+
+def lovasz_hinge_flat(logits: torch.Tensor,
+                      labels: torch.Tensor) -> torch.Tensor:
+    """Binary Lovasz hinge loss.
+
+    Args:
+        logits (torch.Tensor): [P], logits at each prediction
+            (between -infty and +infty).
+        labels (torch.Tensor): [P], binary ground truth labels (0 or 1).
+
+    Returns:
+        torch.Tensor: The calculated loss.
+    """
+    if len(labels) == 0:
+        # only void pixels, the gradients should be 0
+        return logits.sum() * 0.
+    signs = 2. * labels.float() - 1.
+    errors = (1. - logits * signs)
+    errors_sorted, perm = torch.sort(errors, dim=0, descending=True)
+    perm = perm.data
+    gt_sorted = labels[perm]
+    grad = lovasz_grad(gt_sorted)
+    loss = torch.dot(F.relu(errors_sorted), grad)
+    return loss
+
+
+def lovasz_hinge(logits: torch.Tensor,
+                 labels: torch.Tensor,
+                 classes: str or List[int] = 'present',
+                 per_image: bool = False,
+                 class_weight: List[float] = None,
+                 reduction: str = 'mean',
+                 avg_factor: int = None,
+                 ignore_index: int = 255) -> torch.Tensor:
+    """Binary Lovasz hinge loss.
+
+    Args:
+        logits (torch.Tensor): [B, H, W], logits at each pixel
+            (between -infty and +infty).
+        labels (torch.Tensor): [B, H, W], binary ground truth masks (0 or 1).
+        classes (str | list[int], optional): Placeholder, to be consistent with
+            other loss. Default: None.
+        per_image (bool, optional): If per_image is True, compute the loss per
```

Comment:
remove optional,  optional is only used for the parameter whose default value is None